Python/ceval.c

```diff
@@ -863,6 +863,75 @@ static const binaryfunc binary_ops[] = {
     [NB_INPLACE_XOR] = PyNumber_InPlaceXor,
 };
 
+#define NEXT_OP_STORES(O)                                  \
+    ((_Py_OPCODE(*next_instr) == STORE_FAST ||             \
+      _Py_OPCODE(*next_instr) == STORE_FAST__LOAD_FAST) && \
+     GETLOCAL(_Py_OPARG(*next_instr)) == (O))              \
+
+#define BINARY_OP_FAST_INT(OP)                                                 \
+    do {                                                                       \
+        PyObject *lhs = SECOND();                                              \
+        PyObject *rhs = TOP();                                                 \
+        DEOPT_IF(!PyLong_CheckExact(lhs), BINARY_OP);                          \
+        DEOPT_IF(!PyLong_CheckExact(rhs), BINARY_OP);                          \
+        DEOPT_IF(1 < Py_ABS(Py_SIZE(lhs)), BINARY_OP);                         \
+        DEOPT_IF(1 < Py_ABS(Py_SIZE(rhs)), BINARY_OP);                         \
+        STAT_INC(BINARY_OP, hit);                                              \
+        stwodigits l = Py_SIZE(lhs) * ((PyLongObject *)lhs)->ob_digit[0];      \
+        stwodigits r = Py_SIZE(rhs) * ((PyLongObject *)rhs)->ob_digit[0];      \
+        stwodigits i = l OP r;                                                 \
+        Py_DECREF(rhs);                                                        \
+        STACK_SHRINK(1);                                                       \
+        if (-_PY_NSMALLNEGINTS <= i && i < _PY_NSMALLPOSINTS) {                \
+            Py_DECREF(lhs);                                                    \
+            PyLongObject *res = &_PyLong_SMALL_INTS[_PY_NSMALLNEGINTS + i];    \
+            Py_INCREF(res);                                                    \
+            SET_TOP((PyObject *)res);                                          \
+            DISPATCH();                                                        \
+        }                                                                      \
+        bool inplace = NEXT_OP_STORES(lhs);                                    \
+        if (Py_ABS(i) < PyLong_BASE && Py_REFCNT(lhs) == inplace + 1) {        \
+            assert(lhs < (PyObject *)&_PyLong_SMALL_INTS[0] ||                 \
```

Comment:
Isn't this unbounded pointer comparison undefined behavior in C?

Reply:
Ah, yep, nice catch. This was the only way I could think of to assert that we're not mutating a cached small int without looping or calling `PyLong_FromLong` or something.

I'll probably just remove it then, unless anybody can think of a better way of asserting that. Note that `l < -_PY_NSMALLNEGINTS || _PY_NSMALLPOSINTS <= l` [doesn't always mean the same thing](https://bugs.python.org/issue46361) here.

Reply:
Will they be equivalent, once https://bugs.python.org/issue4636 is fixed?

Reply:
> Will they be equivalent, once https://bugs.python.org/issue4636 is fixed?

No, but the fix for bpo-46361 will definitely help. ;)

Ideally, yes. With that fix, all of the tests pass now without triggering the modified assert.

However, I think I'd honestly rather just remove the assert, rather than this being the only place in the codebase where we *rely* on small int caching. If we end up mutating a non-cached copy of a small int, it's fine.